### PR TITLE
Milestone 5: Friendship V1

### PR DESCRIPTION
**Pull request by realisticattorney**
For this milestone I migrate the friendship table, add the model association and methods for checking friendship between users

### CODE REVIEW

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @realisticattorney 
Good job! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed.

##

The project is missing the following behaviour  👇, Please implement the following behaviour.
- [ ]  User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- [ ] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
 - [ ] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
 - [ ] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [ ] There is a button to invite the user to friendship on the user's list page and on the single-user page.([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Comment by kapkev** [comment]
# I wasn't able to review your project due an issue on my machine 
### Kindly ask for another Code Review.
## As always, I wish you all the very best and see you at the top🚀


👨🏻‍💻 **TSE: Kevin**

  
📫 Have any question? Better call Kevin:

[](https://microverse-students.slack.com/team/UT86D3AP4)

**Review by LamiaSristy — CHANGES_REQUESTED** [review]
# Status: Change Required. ♻️
Hello @realisticattorney, Nice work on this. 👍 👍 👍

## The highlights: 
**Thank you for adding the changes. Now those feature is working fine.**
## What to do next:  
Please check my points below and add those required points 👇 

**[REQUIRED]**
- [x] Users will be able to see the “Timeline” page with posts  written themselves and by their friends. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)
Please also check my added images for a better understanding:
Image 1: Here the user hasn't sent an invitation yet ass the buttons are there for sending friend request: 
err-1

But still can see all the posts from the users, who are not friends, even add comments on their posts.
err2


- [x]  In the `README` file, add details of the author of this project.
- [x]  In `README` file in the section`Run tests` there is a typo,  `rpsec --format documentation` it will be rspec.
- [x]  Also, while I am running the test cases, it is falling in my machine. Please check the images 👇 
err3
Please see and correct them.
##

Please make the changes and ask for a re-review.

Feel free to reach me on slack @lamiasristy 🙋

Keep up the spirit always. Happy Coding 💻

File: app/models/user.rb [hunk 1]
```diff
@@ -9,4 +9,38 @@ class User < ApplicationRecord
...
+  def pending_friends
+    friendships.map { |friendship| friendship.friend unless friendship.confirmed }.compact
+  end
+
+  # Users who have requested to be friends
+  def friend_requests
+    inverse_friendships.map { |friendship| friendship.user unless friendship.confirmed }.compact
+  end
+
+  # si la lista friends incluye al usuario es porque es amigo de alguien e.g User.first.friend?(User.find(2))
```
Comment (acushlakoncept) [inline app/models/user.rb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
See this [post](https://stackoverflow.com/questions/13130859/inline-comments-in-ruby/13130900#:~:text=No%2C%20Ruby%20does%20not%20have,comment%20out%20the%20one%20row), about why comments are discouraged.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hi @realisticattorney, 

### You've done an awesome job so far 👌.  
Good job on updating your ReadMe file 💯, adding tests and your apps work just as expected.

But there are still some changes that are necessary for this milestone to be approved.

- [ ] Your commit messages should be capitalized. You can watch this [video](https://www.loom.com/share/00771f369b7b47fc968c019ecf0d3ba6) on how to edit your commit messages using git rebase
 See [Git Commit Best Practices](https://github.com/trein/dev-best-practices/wiki/Git-Commit-Best-Practices#formatting-rules) for more insights

- [ ] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
see the affected file below:
> `layouts/application.html.erb`
**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Comment by realisticattorney** [comment]
So I waited 7 hours to read that I should capitalize my commits and avoid leaving comments, although this is just a step not even the final pull request. About avoiding logic in views, I did. The file you're pointing out was written by microverse, I just pulled it out. You should tell them.

So, pretty much capitalize my commit comments. Is that requirement importantly enough to justify 7 hours of waiting, plus not let me continue working on the next step? Obviously not.

**Review by oloomoses — APPROVED** [review]
# Congratulations, Milestone Approved. :muscle: :tada: :clap: 
**Great job working on the suggested changes from the previous review.
You have passed all the requirements for this milestone :muscle: 
Feel free to merge and close this pull request.**


**I wish you the best in your next project and see you at the top :rocket:** 
#### Happy Coding
_Slack: Oloo Moses_

**Pull request merged**